tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java

```diff
@@ -211,46 +248,81 @@ private Scope(
    * <p>Ops created with this scope will have a control edge from each of the provided controls. All
    * other properties are inherited from the current scope.
    *
+   * <p>Init ops will never be used as control dependencies, they are assumed to be created during
+   * session initialization.
+   *
    * @param controls control dependencies for ops created with the returned scope
    * @return a new scope with the provided control dependencies
    */
   public Scope withControlDependencies(Iterable<Op> controls) {
-    for (Op control : controls) {
+    return withControlDependencyOps(
+        StreamSupport.stream(controls.spliterator(), false)
+            .map(Op::op)
+            .collect(Collectors.toList()));
+  }
+
+  /**
+   * Returns a new scope where added operations will have the provided control dependencies.
+   *
+   * <p>Ops created with this scope will have a control edge from each of the provided controls. All
+   * other properties are inherited from the current scope.
+   *
+   * <p>Init ops will never be used as control dependencies, they are assumed to be created during
+   * session initialization.
+   *
+   * @param controls control dependencies for ops created with the returned scope
+   * @return a new scope with the provided control dependencies
+   */
+  public Scope withControlDependencyOps(Iterable<Operation> controls) {
+    for (Operation control : controls) {
       env.checkInput(control);
     }
-    return new Scope(env, nameScope, controls, deviceSpec);
+
+    return new Scope(env, nameScope, controls, deviceSpec, isInit);
   }
 
   /**
    * Applies device specification and adds each Operand in controlDependencies as a control input to
    * the provided builder.
    *
+   * <p><b>Should only be used from {@link OperationBuilder} implementations</b>
+   *
    * @param builder OperationBuilder to add control inputs and device specification to
    */
   public OperationBuilder apply(OperationBuilder builder) {
     builder.setDevice(deviceSpec.toString());
-    return applyControlDependencies(builder);
+    for (Operation control : controlDependencies) {
+      if (isInit || !env.isInitOp(control)) {
+        builder.addControlInput(control);
+      }
```

Comment:
I have difficulties to understand, so it can occurs that init ops are being added to non-init scopes?

Reply:
This is saying that non-init ops shouldn't have control dependencies on init ops, since they will be done in initialization.  However, init ops should still be able to have control dependencies to other init ops.

Reply:
I guess then that it should not be allowed to do `tf.withControlDependencies(controls)` with init ops in the `controls` list, unless `tf` is within an init scope? Not sure if mixing init and non-init ops in the control list is a valid use case. Can we validate it upfront then and throw an exception in case of incoherence?

Reply:
Yeah.  I don't want to throw an exception if you use an init op as a control dependency, since it's valid semantically, it's just always executed beforehand so we don't need to add anything more.  I will throw if you try to depend on a non-init op in an init scope though.